Detect Multiple Labels capability & ADD-PATH heuristics

ccakes opened the pull request:
Adds detection (not yet NLRI parsing) for multiple labels capability ([RFC 8277](https://tools.ietf.org/html/rfc8277)) and uses heuristics everywhere to detect whether there's a path ID on a given NLRI.

My use case here so far is decoding BMP RouteMonitoring messages which are post-state machine and relying on negotiated capabilities there is a bit hit-and-miss. I realise this probably isn't _really_ the implementation wanted in a crate thats clearly working towards being usable as a BGP speaker so this PR is just to start a discussion about how best to support both use cases

ccakes committed:
Use heuristics to detect path ID in NLRI

ccakes committed:
Detect Multiple Labels capability (RFC8277)

This doesn't actually parse the NLRI yet, just detects the capability

ccakes committed:
fix return type

ccakes committed:
rustfmt

DevQps commented:
Sorry for my late reply on this! The past week has been busy again (like almost every week it seems now...)

I wonder: Is there a way to unambiguously identify when path IDs are used? I can imagine the only way to find this out is by checking the negotiated parameters in the OPEN exchange right? If it cannot be unambiguously identified maybe (to avoid weird bugs) we could add another option to the Capability field which states where or not the stream uses path IDs? 

Of course there are streams which contain multiple peerings of which some may use path IDs and some may not. I guess the only way to fix that is to use different readers/Capabilities to differentiate between those to. Or to try heuristics, but this might of course go wrong when we interpret things wrong...

What do you think?

ccakes commented:
> Sorry for my late reply on this!

No problem! I did actually intend on coming back with some proposed options as well.. 😅

The problem I'm facing is that BMP speakers _should_ send the sent & received OPEN messages as part of the [Peer Up](https://tools.ietf.org/html/rfc7854#section-4.10) notification which I can use to work out the capabilities. some poorly-behaving implementations however (👀 XR 6.4.2...) don't always include them - so I'm left guessing. I can't wait for a parse error then flip the switch either because decoding the MP_REACH_NLRI happens too deep in the codepath to influence.

At a glance, the only options I can think of are:
1. Return more specific errors from `bgp-rs` that can be caught and actioned from other crates. eg check for `bgp_rs::Error::ErrorDecodingPrefixWithoutPathId` then I'd know to flip the switch. This gets hairy though since it'd be a rather large change to describe all the potential errors as enum variants (even nested..)
2. Change the type of [`Capabilities.ADD_PATH_SUPPORT`](https://docs.rs/bgp-rs/0.6.0/bgp_rs/open/struct.Capabilities.html#structfield.ADD_PATH_SUPPORT) to something like:
    ```rust
    enum AddPathSupport {
        Known(HashMap<(AFI, SAFI), AddPathDirection>),
        Unknown
    }
    ```
    then I can externally control whether the detected capabilities are used or if the heuristics are used
3. Pass through some `bool` argument to signal that `util::detect_add_path_prefix` should be used or add a `parse_with_unknown_path_id` method with code duplication (gross)
4. Overbuild the UPDATE parsing in my own crate and duplicate a pile of code (also gross)
5. ??

Number 2 is the easiest to implement with the least changes overall but honestly, I think number 4 is probably the best. If `bgp-rs` is aiming to be a correct BGP library then we don't really want it guessing whether an NLRI has a Path ID.. the correct action to take if the NLRI can't be parsed is to throw an error and then the caller can take whatever action it wants based on that.

Number 1 would be nice as well but I think it's a bit of a tangent to the root issue I'm trying to solve here.

ccakes committed:
Fix bug with capability encoding

and derive Eq for Capabilities and use that in the round-trip tests to
avoid missing bugs in encode

ccakes committed:
Refactor some loops and Path ID detection

DevQps commented:
@ccakes I am currently on vacation but I will be back next friday! Hopefully thats ok! Know however that I really appreciate all the work you put in here.

To give a quick reply!: I think guessing is probably something we should avoid as much as we can. Better and more detailed errors definitely sounds like a potential option. But should probably plan that carefully as well! I hope to take a look once I am back!

ccakes commented:
No problem - I've been on and off vacation days as well. I'll make sure to spend some time on this and have a concrete idea on how to proceed before you're back 😊

ccakes commented:
Closing this - will open a new one in the next couple of days with the heuristics removed 👍

DevQps commented:
That's fine! Thanks! Next Wednesday I have a day off again. I hope I can review it bybthat time! Life here is once again busier then I hoped...

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Cameron Daniel <notifications@github.com>
Sent: Thursday, August 20, 2020 1:14:55 PM
To: DevQps/bgp-rs <bgp-rs@noreply.github.com>
Cc: Christian Veenman <chris_veenman@hotmail.com>; Comment <comment@noreply.github.com>
Subject: Re: [DevQps/bgp-rs] Detect Multiple Labels capability & ADD-PATH heuristics (#41)


Closing this - will open a new one in the next couple of days with the heuristics removed 👍

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/DevQps/bgp-rs/pull/41#issuecomment-677540932>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ALFZIMWNZHKJO6SKT2Q5G7DSBUAS7ANCNFSM4OVW6LIQ>.
